File: packages/marketing-website/scripts/build-sitemap.py
#!/usr/bin/env python3
"""Generate expanded sitemap.xml with 136 URLs (8 langs × all page types)."""
import os
from datetime import date

ROOT = "/Volumes/XPSSD/workspaces/SocialEraser/packages/marketing-website"
BASE = "https://socialeraser.app"
TODAY = "2026-07-07"

# Per-page priority scheme (existing en sitemap is the reference)
# 1.0  = en homepage
# 0.9  = en platform subpages
# 0.8  = non-en homepages + all guide pages
# 0.7  = support
# 0.5  = help + about
# 0.3  = privacy + terms
# 0.2  = success

LANG_PREFIXES = [
    ("",     0),  # en (root)
    ("zh/",  1),
    ("ja/",  1),
    ("es/",  1),
    ("fr/",  1),
    ("de/",  1),
    ("pt/",  1),
    ("ko/",  1),
]

PLATFORMS = ["x", "tiktok", "youtube", "instagram", "facebook"]
GUIDES = ["twitter", "tiktok", "youtube", "instagram", "facebook"]
CONTENT = [
    ("about",    0.5, "monthly"),
    ("help",     0.5, "monthly"),
    ("support",  0.7, "monthly"),
    ("terms",    0.3, "monthly"),
    ("privacy",  0.3, "monthly"),
    ("success",  0.2, "yearly"),
]


def url_block(loc, priority, changefreq, lastmod=TODAY, image=None):
    parts = ["  <url>"]
    parts.append(f"    <loc>{loc}</loc>")
    parts.append(f"    <lastmod>{lastmod}</lastmod>")
    parts.append(f"    <changefreq>{changefreq}</changefreq>")
    parts.append(f"    <priority>{priority}</priority>")
    if image:
        parts.append("    <image:image>")
        parts.append(f"      <image:loc>{image['loc']}</image:loc>")
        parts.append(f"      <image:title>{image['title']}</image:title>")
        parts.append(f"      <image:caption>{image['caption']}</image:caption>")
        parts.append("    </image:image>")
    parts.append("  </url>")
    return "\n".join(parts)


def build():
    blocks = []
    blocks.append('<?xml version="1.0" encoding="UTF-8"?>')
    blocks.append('<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"')
    blocks.append('        xmlns:image="http://www.google.com/schemas/sitemap-image/1.1">')

    # 1) Homepages (8)
    home_images = {
        "en": {
            "loc": f"{BASE}/assets/icons/og-home.png",
            "title": "SocialEraser — Bulk Delete for X, TikTok, YouTube, Instagram &amp; Facebook",
            "caption": "Bulk delete posts, likes, follows, and saved items on 5 social platforms. 100% local, no password."
        }
    }
    for prefix, _ in LANG_PREFIXES:
        if prefix == "":
            loc = f"{BASE}/"
            priority = 1.0
            image = home_images["en"]
        else:
            loc = f"{BASE}/{prefix}"
            priority = 0.8
            image = None
        blocks.append(url_block(loc, priority, "weekly", image=image))

    # 2) Platform subpages (8 langs × 5 = 40)
    for prefix, _ in LANG_PREFIXES:
        for plat in PLATFORMS:
            if prefix == "":
                loc = f"{BASE}/platforms/{plat}/"
            else:
                loc = f"{BASE}/{prefix}platforms/{plat}/"
            priority = 0.9 if prefix == "" else 0.8
            image = None
            if prefix == "" and plat == "x":
                image = {
                    "loc": f"{BASE}/assets/icons/og-x.png",
                    "title": "SocialEraser for X — Bulk Delete Tweets, Likes &amp; Bookmarks",
                    "caption": "Free Chrome extension to bulk delete tweets, retweets, likes, bookmarks, and unfollow on X. 100% local."
                }
            elif prefix == "" and plat == "tiktok":
                image = {
                    "loc": f"{BASE}/assets/icons/og-tiktok.png",
                    "title": "SocialEraser for TikTok — Bulk Delete Videos, Reposts &amp; Likes",
                    "caption": "Free Chrome extension to bulk delete TikTok videos, reposts, likes, favorites, and unfollow. 100% local."
                }
            blocks.append(url_block(loc, priority, "weekly", image=image))

    # 3) Guide pages (8 langs × 5 = 40)
    for prefix, _ in LANG_PREFIXES:
        for guide in GUIDES:
            if prefix == "":
                loc = f"{BASE}/guides/{guide}.html"
            else:
                loc = f"{BASE}/{prefix}guides/{guide}.html"
            priority = 0.8
            blocks.append(url_block(loc, priority, "monthly"))

    # 4) Content pages (8 langs × 6 = 48)
    for prefix, _ in LANG_PREFIXES:
        for name, priority, freq in CONTENT:
            if prefix == "":
                loc = f"{BASE}/{name}.html"
            else:
                loc = f"{BASE}/{prefix}{name}.html"
            blocks.append(url_block(loc, priority, freq))

    blocks.append("</urlset>")
    return "\n".join(blocks) + "\n"


if __name__ == "__main__":
    out = os.path.join(ROOT, "sitemap.xml")
    with open(out, "w", encoding="utf-8") as f:
        f.write(build())
    # Count
    count = build().count("<url>")
    print(f"Wrote {out} with {count} <url> blocks")

File: packages/marketing-website/scripts/build-sitemap.js
#!/usr/bin/env node
// One-shot: rebuild sitemap.xml from the actual filesystem so URL formats
// stay in lockstep with the live content.
//
// Behaviour:
//   - Globs every *.html under the marketing-website root (skips scripts/,
//     node_modules/, .git/).
//   - Computes each page's URL: pages served by <path>/index.html get a
//     trailing slash; pages served by a flat *.html file get the path
//     without .html (e.g. /about, /guides/twitter).
//   - Sets <lastmod> to a fixed release date (LASTMOD_DATE). We pin every
//     URL to the same date so the sitemap reflects the date the site went
//     live with the unified /path form, not the mtime of whichever file
//     was touched last in CI. Update LASTMOD_DATE only when the site
//     actually changes in a way that should move every page's freshness
//     signal — for incremental updates, prefer a new sitemap submission
//     without bumping the global date.
//   - For platform subpages, attaches an <image:image> block:
//       - X and TikTok: dedicated OG image (og-x.png / og-tiktok.png)
//       - YouTube / Instagram / Facebook: og-home.png placeholder
//         (matches the og:image the HTML actually points at now).
//   - Home page gets the og-home.png image too.
//
// Run: `node scripts/build-sitemap.js`

const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..');
const SITE = 'https://socialeraser.app';
// Pin every <lastmod> to a single release date so the sitemap's freshness
// signal matches the date the unified /path URL form went live, regardless
// of which files were touched by the build. Bump LASTMOD_DATE whenever the
// site has a release that should reset the freshness signal for all URLs.
const LASTMOD_DATE = '2026-07-07';

const SKIP_DIRS = new Set(['node_modules', 'scripts', '.git']);
const PRIORITY_HOME = '1.0';
const PRIORITY_PLATFORM = '0.9';
const PRIORITY_LANG_HOME = '0.8';
const PRIORITY_LANG_PLATFORM = '0.8';
const PRIORITY_GUIDE = '0.8';
const PRIORITY_SUPPORT = '0.7';
const PRIORITY_ABOUT_HELP = '0.5';
const PRIORITY_LEGAL = '0.3';
const PRIORITY_SUCCESS = '0.2';

function walk(dir) {
  const out = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    if (SKIP_DIRS.has(entry.name)) continue;
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) out.push(...walk(full));
    else if (entry.isFile() && entry.name.endsWith('.html')) out.push(full);
  }
  return out;
}

function pageUrlFor(file) {
  const rel = path.relative(ROOT, file).replace(/\\/g, '/');
  if (rel === 'index.html') return '/';
  // Strip a trailing /index.html → directory URL with trailing slash
  if (rel.endsWith('/index.html')) {
    return '/' + rel.slice(0, -'/index.html'.length) + '/';
  }
  // Flat *.html → path without extension
  return '/' + rel.replace(/\.html$/, '');
}

function ymd(_mtime) {
  return LASTMOD_DATE;
}

function classify(url) {
  if (url === '/') return { priority: PRIORITY_HOME, changefreq: 'weekly', image: 'og-home.png' };
  if (/^\/(zh|ja|es|fr|de|pt|ko)\/$/.test(url)) return { priority: PRIORITY_LANG_HOME, changefreq: 'weekly' };
  if (/^\/platforms\/[a-z]+\/$/.test(url)) {
    const plat = url.split('/')[2];
    const platformImages = { x: 'og-x.png', tiktok: 'og-tiktok.png' };
    return { priority: PRIORITY_PLATFORM, changefreq: 'weekly', image: platformImages[plat] || 'og-home.png' };
  }
  if (/^\/(zh|ja|es|fr|de|pt|ko)\/platforms\/[a-z]+\/$/.test(url)) {
    const plat = url.split('/')[3];
    const platformImages = { x: 'og-x.png', tiktok: 'og-tiktok.png' };
    return { priority: PRIORITY_LANG_PLATFORM, changefreq: 'weekly', image: platformImages[plat] || 'og-home.png' };
  }
  if (/^\/(guides|zh\/guides|ja\/guides|es\/guides|fr\/guides|de\/guides|pt\/guides|ko\/guides)\/[a-z]+$/.test(url))
    return { priority: PRIORITY_GUIDE, changefreq: 'monthly' };
  if (/^\/(support|zh\/support|ja\/support|es\/support|fr\/support|de\/support|pt\/support|ko\/support)$/.test(url))
    return { priority: PRIORITY_SUPPORT, changefreq: 'monthly' };
  if (/^\/(about|help|zh\/about|zh\/help|ja\/about|ja\/help|es\/about|es\/help|fr\/about|fr\/help|de\/about|de\/help|pt\/about|pt\/help|ko\/about|ko\/help)$/.test(url))
    return { priority: PRIORITY_ABOUT_HELP, changefreq: 'monthly' };
  if (/^\/(terms|privacy|zh\/terms|zh\/privacy|ja\/terms|ja\/privacy|es\/terms|es\/privacy|fr\/terms|fr\/privacy|de\/terms|de\/privacy|pt\/terms|pt\/privacy|ko\/terms|ko\/privacy)$/.test(url))
    return { priority: PRIORITY_LEGAL, changefreq: 'monthly' };
  if (/^\/(success|zh\/success|ja\/success|es\/success|fr\/success|de\/success|pt\/success|ko\/success)$/.test(url))
    return { priority: PRIORITY_SUCCESS, changefreq: 'yearly' };
  return { priority: '0.5', changefreq: 'monthly' };
}

const files = walk(ROOT);
// 404 is a noindex page (see <meta name="robots" content="noindex, follow">)
// and a duplicate of an HTTP 404 response — it should not appear in the
// sitemap. Excluding it here keeps the sitemap's URL set clean and avoids
// sending crawlers a path that explicitly asks them not to index.
const realFiles = files.filter(f => {
  const rel = path.relative(ROOT, f).replace(/\\/g, '/');
  return rel !== '404.html';
});
const entries = realFiles.map(f => {
  const url = pageUrlFor(f);
  const stat = fs.statSync(f);
  return { url, mtime: stat.mtime, ...classify(url) };
});

// Sort: home first, then language homepages, then by URL for stable ordering.
entries.sort((a, b) => a.url.localeCompare(b.url));

const out = [];
out.push('<?xml version="1.0" encoding="UTF-8"?>');
out.push('<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"');
out.push('        xmlns:image="http://www.google.com/schemas/sitemap-image/1.1">');

for (const e of entries) {
  out.push('  <url>');
  out.push(`    <loc>${SITE}${e.url}</loc>`);
  out.push(`    <lastmod>${ymd(e.mtime)}</lastmod>`);
  out.push(`    <changefreq>${e.changefreq}</changefreq>`);
  out.push(`    <priority>${e.priority}</priority>`);
  if (e.image) {
    out.push('    <image:image>');
    out.push(`      <image:loc>${SITE}/assets/icons/${e.image}</image:loc>`);
    out.push('    </image:image>');
  }
  out.push('  </url>');
}
out.push('</urlset>');
out.push('');

const sitemapPath = path.join(ROOT, 'sitemap.xml');
fs.writeFileSync(sitemapPath, out.join('\n'));
console.log(`build-sitemap: ${entries.length} URLs written to sitemap.xml`);
